### PR TITLE
Model manager

### PR DESCRIPTION
**Pull request by kecso**
This simple visualizer allows easy management for models inside a single webgme project.
The visualizer is mainly intended for domain-specific use as it requires proper configuration in the
knowledge of the domain.
Design studios could leverage its easy access.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,6 +22,7 @@
     "chai": "^3.0.0"
   },
   "dependencies": {
-    "jszip": "^3.1.3"
+    "jszip": "^3.1.3",
+    "webgme": "^2.25.0"
```
Comment (pmeijer) [inline package.json]: webgme should be in peerDependencies

File: README.md [hunk 2]
```diff
@@ -13,6 +13,23 @@ Note that this visualizer uses `eval` to evaluate the code and is recommended to
...
 ![ICore](images/icore.png "ICore in action - note the controls in the toolbar")
 
+#### ModelManager
+Configurable, domain-specific, easy-to-use model management visualizer. It allows the creation, deletion
+and import/export of models. Note, that the component configuration has to be set properly to work!
+```
+{...
+    "ModelManager":
+    {
+        "container": "pathOfTheContainer",
```
Comment (pmeijer) [inline README.md:23]: Is the root default here?

File: README.md [hunk 3]
```diff
@@ -13,6 +13,23 @@ Note that this visualizer uses `eval` to evaluate the code and is recommended to
...
+Configurable, domain-specific, easy-to-use model management visualizer. It allows the creation, deletion
+and import/export of models. Note, that the component configuration has to be set properly to work!
+```
+{...
+    "ModelManager":
+    {
+        "container": "pathOfTheContainer", 
+        /* the listed models will be the children of this 
+        node and new ones will be created here */
+        "types": ['componentName','otherComponentName']
```
Comment (pmeijer) [inline README.md:26]: It looks like the part-browser follows this as well - is that the case?

File: src/visualizers/widgets/ModelManager/ModelManagerWidget.js [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+/**
+ * Generated by VisualizerGenerator 1.7.0 from webgme on Thu Apr 12 2018 12:02:53 GMT-0500 (CDT).
+ */
+
+define(['css!./styles/ModelManagerWidget.css'], function () {
+    'use strict';
+
+    var ModelManagerWidget,
+        WIDGET_CLASS = 'model-manager',
+        BASE_WIDGET_FRAME = '<div class="col-sm-12">\n' +
```
Comment (pmeijer) [inline src/visualizers/widgets/ModelManager/ModelManagerWidget.js]: For future reference you can use the new string templates, using `, for multi-line strings.

File: src/visualizers/widgets/ModelManager/ModelManagerWidget.js [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+            '    <div id="toolbarBtns" class="col-sm-offset-2">\n' +
+            '    </div>\n' +
+            '    <div class="col-sm-12">\n' +
+            '      <table class="table"><thead><tr><th>Name</th><th>Type</th><th></th></tr></thead>\n' +
+            '        <tbody id="modelList"></tbody>\n' +
+            '      </table>\n' +
+            '    </div>\n' +
+            '</div>',
+        SINGLE_CREATE_BUTTON = '<button id="createBtn" type="button" class="btn btn-danger">Create a new</button>',
```
Comment (pmeijer) [inline src/visualizers/widgets/ModelManager/ModelManagerWidget.js]: This should probably not be danger (red).

File: src/visualizers/widgets/ModelManager/ModelManagerWidget.js [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+                paths.forEach(function (metaPath) {
+                    if (self._config.types.indexOf(core.getAttribute(metaNodes[metaPath], 'name')) !== -1) {
+                        missing -= 1;
+                        self._types[core.getAttribute(metaNodes[metaPath], 'name')] = metaPath;
+                    }
+                });
+
+                if (missing !== 0) {
+                    self._logger.warn('Incompatible component setting:', self._config);
+                    self._config.inFault = true;
```
Comment (pmeijer) [inline src/visualizers/widgets/ModelManager/ModelManagerWidget.js]: If one is missing, can it still proceed?

**Review by pmeijer — CHANGES_REQUESTED** [review]
I have some comments/questions.
Btw why not having a dedicated repo? That way it can show up on webgme.org

File: README.md (hunk 2 above)
Comment (kecso) [reply README.md:23]: no, right now there is no default anywhere and the container of choice has to exsist for the visualizer to work

File: README.md (hunk 3 above)
Comment (kecso) [reply README.md:26]: no, that is just a coincidence of the picture.
this visualizer has nothing to do with the part browser though I believe a good configuration will match in that regard

File: src/visualizers/widgets/ModelManager/ModelManagerWidget.js (hunk 4 above)
Comment (kecso) [reply src/visualizers/widgets/ModelManager/ModelManagerWidget.js]: I'll change it

File: src/visualizers/widgets/ModelManager/ModelManagerWidget.js (hunk 5 above)
Comment (kecso) [reply src/visualizers/widgets/ModelManager/ModelManagerWidget.js]: changing it to default or something then

File: src/visualizers/widgets/ModelManager/ModelManagerWidget.js (hunk 6 above)
Comment (kecso) [reply src/visualizers/widgets/ModelManager/ModelManagerWidget.js]: right now not.
it is rigid in that sense, but I wanted it that way to make it more tailored for design studio use

**Comment by kecso** [comment]
I thought having it as a 'generic' visualizer makes it more visible.
But you might be right (I think we need to advertise it anyhow as part of a design studio know how...)

File: README.md (hunk 3 above)
Comment (pmeijer) [reply README.md:26]: Why not hide the parts in the partbrowser then? Would be esp. nice if you're using the root-node as the container to not show all the parts. And considering that there's been requests to hide the parts of the root-node using this as a "root-viz" would solve that issue.

**Comment by pmeijer** [comment]
Testing it now and it throws an exception if there is no config defined.
I'd suggest using the rootNode as default container and if no types are defined just display all.
```
q.js:349 Uncaught TypeError: Cannot read property 'length' of undefined
    at ModelManagerWidget.js:76
    at q.js:2243
    at runSingle (q.js:325)
    at MessagePort.flush (q.js:313)
```

File: src/visualizers/widgets/ModelManager/ModelManagerWidget.js [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+            elementStr += model.type;
+            elementStr += '</td><td>';
+            elementStr += '<button type="button" id="' + index + '_viewBtn" class="btn btn-primary">View</button>';
+            elementStr += '<button type="button" id="' + index + '_exportBtn" class="btn btn-info">Export</button>';
+            elementStr += '<button type="button" id="' + index + '_deleteBtn" class="btn btn-danger">Delete</button>';
+            elementStr += '</td></tr>';
+
+            $(self._list).append(elementStr);
+            $(self._list).find('#' + index + '_viewBtn').click(function () {
+                console.log('yeah, I am going to view :', model.path);
```
Comment (pmeijer) [inline src/visualizers/widgets/ModelManager/ModelManagerWidget.js]: Left out logs

**Comment by pmeijer** [comment]
Would be nice if it filtered the existing children based on the provided types

**Comment by pmeijer** [comment]
Also consider pulling the "Create new.." and "Import buttons" to the right or left.

**Review by pmeijer — APPROVED** [review]

**Pull request merged**